src/main/drivers/accgyro/accgyro_spi_icm456xx.c

```diff
@@ -0,0 +1,641 @@
+        break;
+    default:
+        gyro->scale = GYRO_SCALE_2000DPS;
+        gyro->gyroRateKHz = GYRO_RATE_6400_Hz;
+        gyro->gyroSampleRateHz = 6400;
+        spiWriteReg(dev, ICM456XX_GYRO_CONFIG0, ICM456XX_GYRO_FS_SEL_2000DPS | ICM456XX_GYRO_ODR_6K4_LN);
+        break;
+    }
+
+    gyro->gyroShortPeriod = clockMicrosToCycles(HZ_TO_US(gyro->gyroSampleRateHz));
+
+    spiWriteReg(dev, ICM456XX_INT1_CONFIG2, ICM456XX_INT1_MODE_PULSED | ICM456XX_INT1_DRIVE_CIRCUIT_PP |
+                                            ICM456XX_INT1_POLARITY_ACTIVE_HIGH);
+
+    spiWriteReg(dev, ICM456XX_INT1_CONFIG0, ICM456XX_INT1_STATUS_EN_DRDY);
+    
+}
+ 
+uint8_t icm456xxSpiDetect(const extDevice_t *dev)
+{
+    uint8_t icmDetected = MPU_NONE;
+    uint8_t attemptsRemaining = 20;
+    uint32_t waited_us = 0;
+
+    // Soft reset
+    spiWriteReg(dev, ICM456XX_REG_MISC2, ICM456XX_SOFT_RESET);
+
+    // Wait for reset to complete (bit 1 of REG_MISC2 becomes 0)
+    do {
+        uint8_t misc2 = spiReadRegMsk(dev, ICM456XX_REG_MISC2);
+        if ((misc2 & ICM456XX_SOFT_RESET) == 0) {
+            break; // Reset complete
+        }
+        delayMicroseconds(10);
+        waited_us += 10;
+
+        if (waited_us >= ICM456XX_RESET_TIMEOUT_US) {
+            return MPU_NONE;
+        }
+
+    } while (true);
+
+    spiWriteReg(dev, ICM456XX_REG_BANK_SEL, ICM456XX_BANK_0);
+
+    do {
+        const uint8_t whoAmI = spiReadRegMsk(dev, ICM456XX_WHO_AM_REGISTER);
+        switch (whoAmI) {
+        case ICM45686_WHO_AM_I_CONST:
+            icmDetected = ICM_45686_SPI;
+            break;
+        case ICM45605_WHO_AM_I_CONST:    
+            icmDetected = ICM_45605_SPI;
+            break;
+        default:
+            icmDetected = MPU_NONE;
+            break;
+        }
+
+    } while (icmDetected == MPU_NONE && attemptsRemaining--);
+
+    return icmDetected;
+
+}
+
+bool icm456xxAccReadSPI(accDev_t *acc)
+{
+    switch (acc->gyro->gyroModeSPI) {
+    case GYRO_EXTI_INT:
+    case GYRO_EXTI_NO_INT:
+    {
+#ifdef USE_DMA
+        if (spiUseDMA(&acc->gyro->dev)) {      
+            acc->gyro->dev.txBuf[0] = ICM456XX_ACCEL_DATA_X1_UI | 0x80;
+
+            busSegment_t segments[] = {
+                    {.u.buffers = {NULL, NULL}, ICM456XX_SPI_BUFFER_SIZE, true, NULL},
+                    {.u.link    = {NULL, NULL}, 0, true, NULL},
+            };
+            memset(&acc->gyro->dev.txBuf[1], 0xFF, 6);
+            segments[0].u.buffers.txData = acc->gyro->dev.txBuf;
+            segments[0].u.buffers.rxData = &acc->gyro->dev.rxBuf[1];
+            spiSequence(&acc->gyro->dev, &segments[0]);
+
+            // Wait for completion
+            spiWait(&acc->gyro->dev);
+
+            // Fall through
+            FALLTHROUGH;
+        } else 
+#else
+        {
+           // Interrupts are present, but no DMA. Non-DMA read
+           uint8_t raw[ICM456XX_DATA_LENGTH];
```

Comment:
(you can probably use  &acc->gyro->dev.rxBuf[1]; here too, but check that it is available even without DMA)

Reply:
hm, need to check

Reply:
not work